opendevin/controller/agent_controller.py

```diff
@@ -319,13 +319,13 @@ async def _step(self):
         else:
             await self.add_history(action, NullObservation(''))
 
+        if not isinstance(action, NullAction):
+            await self.event_stream.add_event(action, EventSource.AGENT)
```

Comment:
This would cause some trouble for delegation at least. When you add an event to the stream, its callback is called. In delegation scenario, the control is then handed over to child agent (controller). Even if we just exceeded the budget during this step, the child agent would still continue because we haven't computed and checked budget yet.

Reply:
Is this scenario you're thinking of at the beginning of delegation, so when the event is a delegate action?

Reply:
Yes, when event is `AgentDelegateAction`

Reply:
Okay, we could stop it at some point ( perhaps `on_event` or follow up), though I'm not sure it will be pretty. Or when the controller is about to start and initializes the budget (oops error). I guess we could even, if it seems better, set the agent to ERROR, then add the event; to check for that state and stop the process. I'm not anymore at the computer, will give it some thought. If we can't, IMHO an alternative is to cancel, that is, this action "doesn't exist". I still don't see it in history unless it's in the stream...

The funny thing here is that I literally couldn't merge the current logic. Because histories are copies of the event stream in `main`, and the copies will vanish when we refactor to have the event stream as the source of truth. There's no 'space' between add_history and add_event anymore 😅

Reply:
In the worst case, we could merge this PR as it is, and don't care about the budget thing I mentioned. It won't be ideal but not too bad.

Reply:
I would agree, but maybe the simplest is to just move the budget check before the call to the agent `step`. It seems like it would work, to stop the next step... Will check.

Reply:
> but maybe the simplest is to just move the budget check before the call to the agent step

Oh yeah sounds like a plan.